### ISSUE TITLE
about `predictor_proportion`

### ISSUE THREAD

**Issue by xixiha5230** [alirezakazemipour/PPO-RND#1]
Thanks for the open source code, I'm currently having a bit of a problem with `predictor_proportion`.

What is the use of this, why use this to generate rnd loss mask？

**Comment by akazemipour** [comment]
Hi @xixiha5230 
Thank you for your inquiry!
These two sections in the paper answer your question:
- > To hold the rate at which the intrinsic reward decreases over time constant across experiments with
different numbers of parallel environments, we downsample the batch size when training the predictor
to match the batch size with 32 parallel environments (for full details see Appendix A.4). Larger
numbers of environments results in larger batch sizes per update for training the policy, whereas
the predictor network batch size remains constant. Since the intrinsic reward disappears over time
it is important for the policy to learn to find and exploit these transitory rewards, since they act as
stepping-stones to nearby novel states.

- > Initial preliminary experiments with RND were run with only 32 parallel environments. We expected
that increasing the number of parallel environments would improve performance by allowing the
policy to adapt more quickly to transient intrinsic rewards. This effect could have been mitigated
however if the predictor network also learned more quickly. To avoid this situation when scaling
up from 32 to 128 environments we kept the effective batch size for the predictor network the
same by randomly dropping out elements of the batch with keep probability 0.25. Similarly in
our experiments with 256 and 1,024 environments we dropped experience for the predictor with
respective probabilities 0.125 and 0.03125.

**Comment by xixiha5230** [comment]
Thank you for your response. I understand the meaning of the variable and have implemented RND in my own code framework based on your example. However, I've noticed that the prediction network of RND converges too quickly (my predictor_proportion > 1, but hardware limitations mean that I only have access to 6 workers). Additionally, there doesn't seem to be any intrinsic reward for agent exploration. Are there any suggestions or input you can provide to help with this issue?

This is the external reward, the internal reward, and the loss of the rnd prediction network, and the abscissa represents an update

![image](https://user-images.githubusercontent.com/31526189/231205292-f3aaeba7-2285-4024-bc19-3d50d23a21c4.png)

**Comment by akazemipour** [comment]
> However, I've noticed that the prediction network of RND converges too quickly (my predictor_proportion > 1, but hardware limitations mean that I only have access to 6 workers)

What's your environment? If it isn't very diverse then it makes sense that rnd loss converges quickly to zero as there is not much interesting things to be surprised. Also, What do you mean by converging to zero? even values around magnitudes of 10e-4 aren't that much small. :sweat_smile: 

> Additionally, there doesn't seem to be any intrinsic reward for agent exploration. Are there any suggestions or input you can provide to help with this issue?

It'd be useful if you plot the entropy over your actions as well. As far as the entropy possessing high values, it's okay since intrinsic reward spikes when something surprising happens to the agent rather than being driven by the agent itself. So in short, the agent explores based on entropy and if something surprising happens then it learns from that to visit it more.

> This is the external reward, the internal reward, and the loss of the rnd prediction network, and the abscissa represents an update

A LOT more iterations are needed in order of magnitudes!

**Comment by xixiha5230** [comment]
**Thank you very much for your thoughtful answer !!!**

> What's your environment?

There are two environments for my training，they were trained together. The first one is reward sparse. This car needs to find the football first (get 10 rewards), then a door will open, and then go out through the opened door (get 10 rewards), the task is Completed, the screenshot of the first environment is as follows

![1681263408005](https://user-images.githubusercontent.com/31526189/231324841-7f4bcf72-f4c6-46a9-96c8-ed9fd46cf82a.jpg)

The second environment is reward-dense. The car enters from the bottom door, moves upward and needs to avoid obstacles moving left and right, and then exits from the top door. There will be checkpoints along the way to provide rewards (1 / checkpoint) for the car  , the screenshot of the environment is as follows:
![1681264066747](https://user-images.githubusercontent.com/31526189/231326250-6322e877-4b8b-42b9-b16f-8da19ea52212.jpg)

> If it isn't very diverse then it makes sense that rnd loss converges quickly to zero as there is not much interesting things to be surprised.

The objects in the environment are indeed not very diverse, but I randomized the background of the camera (observation, shape is (84, 84, 3)), sometimes it is a random solid color, sometimes it is Gaussian noise, rnd target network‘s output is a 512-dimensional feature, I thought it would not converge so quickly.

> Also, What do you mean by converging to zero? even values around magnitudes of 10e-4 aren't that much small. 😅

Maybe I didn't phrase it clearly enough, the loss didn't drop to 0, it just dropped too fast (I think)

> It'd be useful if you plot the entropy over your actions as well. As far as the entropy possessing high values, it's okay since intrinsic reward spikes when something surprising happens to the agent rather than being driven by the agent itself. So in short, the agent explores based on entropy and if something surprising happens then it learns from that to visit it more.

After one night of training, the current curve is shown in the figure below. I also added the entropy curve. ~~Although it seems that the intrinsic reward does not play a big role.~~ I forgot to mention that I have done reward scaling for the external reward. I have not drawn the reward after scaling. The reward in this picture has not been scaled, so this should explain why the internal reward is not large, but it works. I observed that the agent performed better than before (without rnd). The task completion rate is higher.

![image](https://user-images.githubusercontent.com/31526189/231328316-ecb4e95b-cdf3-41a0-ac0e-957eb6477215.png)

Thanks again for your thoughtful response!
